### PR TITLE
Bulk ACL management for AsynchronousOperations Admin UI

### LINKED ISSUE #29757: [Issue] Bulk ACL management for AsynchronousOperations Admin UI
This issue is automatically created based on existing pull request: magento/magento2#27580: Bulk ACL management for AsynchronousOperations Admin UI

---------
### Description (*)
After Migrating of Asynchronous Operations from Magento Commerce to Magento Open Source, looks like part of functionality was extended.

In details:
In magento_bulk table was added user_type, which defines type of the user who created Bulk Operation.

Possible types are:
- Admin
- Integration
- Guest
- Customer

In current implementation all Admin UI components have no idea about user type:

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/AsynchronousOperations/view/adminhtml/ui_component/bulk_listing.xml - in default Grid there are NO DataSource is defined, so Admin see the whole operations, but at the same time, he cannot see Details of those operations:

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/AsynchronousOperations/Controller/Adminhtml/Bulk/Details.php#L52

But at you can see from implementation, 
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/AsynchronousOperations/Model/AccessValidator.php#L58

that permissions are checked based on UserID and fully ignoring UserType. Which means, that Admin has access to All transactions or all user types with the same ID.

### Fixed Issues (if relevant)
Current implementation will add: 
* New ACL roles for give possibilities for Admin define permissions for Admin/Integration user to have access only to specific user types operations.
* With restricted user role, Admin will be able to see only operations that are assigned to him, View detailes or Restart them
* Also Admin notifications will be restricted only to allowed.

### Questions or comments
Auto tests still in process, but main implementation can be already reviewed.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by nuzil**
### Description (*)
After Migrating of Asynchronous Operations from Magento Commerce to Magento Open Source, looks like part of functionality was extended.

In details:
In magento_bulk table was added user_type, which defines type of the user who created Bulk Operation.

Possible types are:
- Admin
- Integration
- Guest
- Customer

In current implementation all Admin UI components have no idea about user type:

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/AsynchronousOperations/view/adminhtml/ui_component/bulk_listing.xml - in default Grid there are NO DataSource is defined, so Admin see the whole operations, but at the same time, he cannot see Details of those operations:

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/AsynchronousOperations/Controller/Adminhtml/Bulk/Details.php#L52

But at you can see from implementation, 
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/AsynchronousOperations/Model/AccessValidator.php#L58

that permissions are checked based on UserID and fully ignoring UserType. Which means, that Admin has access to All transactions or all user types with the same ID.

### Fixed Issues (if relevant)
Current implementation will add: 
* New ACL roles for give possibilities for Admin define permissions for Admin/Integration user to have access only to specific user types operations.
* With restricted user role, Admin will be able to see only operations that are assigned to him, View detailes or Restart them
* Also Admin notifications will be restricted only to allowed.

### Questions or comments
Auto tests still in process, but main implementation can be already reviewed.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29757: Bulk ACL management for AsynchronousOperations Admin UI

### CODE REVIEW

**Comment by sidolov** [comment]
@magento create issue

File: lib/internal/Magento/Framework/Bulk/BulkStatusInterface.php [hunk 1]
```diff
@@ -38,9 +38,20 @@ public function getOperationsCountByBulkIdAndStatus($bulkUuid, $status);
...
 
+    /**
+     * Get all bulks created by user and user type
+     *
+     * @param int $userId
+     * @param int $userTypeId
+     * @return BulkSummaryInterface[]
+     * @since 100.2.0
+     */
+    public function getBulksByUserAndType($userId, $userTypeId);
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Bulk/BulkStatusInterface.php:53]: Adding the new methods to the interface is backward-incompatible change. Please, introduce new interface with this method.

File: app/code/Magento/AsynchronousOperations/Model/BulkStatus.php [hunk 2]
```diff
@@ -118,6 +118,31 @@ public function getBulksByUser($userId)
         return $collection->getItems();
     }
 
+    /**
+     * @inheritDoc
+     */
+    public function getBulksByUserAndType($userId, $userTypeId)
```
Comment (sidolov) [inline app/code/Magento/AsynchronousOperations/Model/BulkStatus.php:140]: Should be extracted to the separate class

File: app/code/Magento/AsynchronousOperations/Model/BulkOperationsStatus.php [hunk 3]
```diff
@@ -93,6 +93,14 @@ public function getBulksByUser($userId)
         return $this->bulkStatus->getBulksByUser($userId);
     }
 
+    /**
+     * @inheritDoc
+     */
+    public function getBulksByUserAndType($userId, $userTypeId)
```
Comment (sidolov) [inline app/code/Magento/AsynchronousOperations/Model/BulkOperationsStatus.php:101]: Should be extracted to separate class

File: app/code/Magento/AsynchronousOperations/Model/AccessManager.php [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+use Magento\Framework\AuthorizationInterface;
+use Magento\Framework\EntityManager\EntityManager;
+
+/**
+ * Class AccessManager providing information
+ * about valid resources for Bulk Reports
+ */
+class AccessManager
+{
+    public const BULK_LOGGING_ACL_GUESTS
```
Comment (sidolov) [inline app/code/Magento/AsynchronousOperations/Model/AccessManager.php:20]: Is it possible to make constants private?

File: app/code/Magento/AsynchronousOperations/Model/AccessManager.php [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        $this->allowedUserTypes = $this->getGlobalAllowedUserTypes();
+    }
+
+    /**
+     * Check if content allowed for current use depends from assigned user roles and bulkUuid
+     *
+     * @param int $bulkUuid
+     * @return bool
+     */
+    public function isAllowedForBulkUuid($bulkUuid)
```
Comment (sidolov) [inline app/code/Magento/AsynchronousOperations/Model/AccessManager.php:83]: I prefer to split this class into three with one method.

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by engcom-Charlie** [comment]
@nuzil can you please look at the requested changes?
Thank you.

**Comment by engcom-Charlie** [comment]
Hi @nuzil, I'm closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for your collaboration.

**Comment by nuzil** [comment]
Hi @engcom-Charlie 
I'm on it. As you can see I sync up branches only 4 days ago. Its require some time to refactor so many stuff.

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Moving the PR to the correct state, as @nuzil mentioned, it is being worked on.

**Comment by engcom-Charlie** [comment]
Hi, @nuzil will you continue on this PR?
Thank you.

**Comment by engcom-Charlie** [comment]
Hi @nuzil, I'm closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for your collaboration.

**Comment by engcom-Charlie** [comment]
Closed as duplicate https://github.com/magento/magento2/pull/30806.

**Pull request closed without merging**